pkg/scheduler/reconcile_on_edgeplacement.go

```diff
@@ -21,79 +21,140 @@ import (
 
 	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/klog/v2"
 
 	kcpcache "github.com/kcp-dev/apimachinery/v2/pkg/cache"
+	"github.com/kcp-dev/logicalcluster/v3"
 
 	edgev1alpha1 "github.com/kcp-dev/edge-mc/pkg/apis/edge/v1alpha1"
 )
 
 func (c *controller) reconcileOnEdgePlacement(ctx context.Context, epKey string) error {
 	logger := klog.FromContext(ctx)
-	ws, _, name, err := kcpcache.SplitMetaClusterNamespaceKey(epKey)
+	epws, _, epName, err := kcpcache.SplitMetaClusterNamespaceKey(epKey)
 	if err != nil {
 		logger.Error(err, "invalid EdgePlacement key")
 		return err
 	}
-	logger = logger.WithValues("workspace", ws, "edgePlacement", name)
+	logger = logger.WithValues("workspace", epws, "edgePlacement", epName)
 	ctx = klog.NewContext(ctx, logger)
 	logger.V(2).Info("reconciling")
 
-	// TODO(waltforme): should I use a client to bother the apiserver or use local store?
-	ep, err := c.edgeClusterClient.EdgeV1alpha1().EdgePlacements().Cluster(ws.Path()).Get(ctx, name, metav1.GetOptions{})
+	/*
+		On EdgePlacement 'ep' change:
+		1) from cache, find loc(s) that being selected by ep
+
+		2) from cache, for each of the found loc, find st(s) that being selected by the loc
+
+		3) update store, with loc(s) that being selected by ep
+
+		4) update apiserver
+
+		Need data structure: none.
+	*/
+
+	store.l.Lock()
+	defer store.l.Unlock() // TODO(waltforme): Is it safe to shorten the critical section?
+
+	ep, err := c.edgePlacementLister.Cluster(epws).Get(epName)
 	if err != nil {
 		if errors.IsNotFound(err) {
+			logger.V(1).Info("EdgePlacement not found")
+			logger.V(3).Info("dropping EdgePlacement from store")
+			store.dropEp(epKey)
 			return nil
+		} else {
+			logger.Error(err, "failed to get EdgePlacement")
+			return err
 		}
+	}
+
+	// 1)
+	locsAll, err := c.locationLister.List(labels.Everything())
+	if err != nil {
+		logger.Error(err, "failed to list Locations in all workspaces")
 		return err
 	}
+	locsFilteredByEp, err := filterLocsByEp(locsAll, ep)
+	if err != nil {
+		logger.Error(err, "failed to find Locations for EdgePlacement")
+	}
+	locsSelecting := packLocKeys(locsFilteredByEp)
+
+	singles := []edgev1alpha1.SinglePlacement{}
+	for _, loc := range locsFilteredByEp {
+		// 2)
+		lws := logicalcluster.From(loc)
+		stsInLws, err := c.synctargetLister.Cluster(lws).List(labels.Everything())
```

Comment:
If the store maintained a map from Location[ExternalName] to set of SyncTarget[NameAndUID] then it would not be necessary to enumerate every SyncTarget and evaluate the Location's "where predicate" here.  In the current revision, the number of "where predicate" evaluations is O( (number of Locations selected by this EdgePlacement) X (number of SyncTargets in the world) ).